shotover/src/codec/cassandra.rs

```diff
@@ -221,43 +219,75 @@ impl CassandraDecoder {
+                    if header_crc24 != computed_crc {
+                        return Err(header_crc_mismatch_error(computed_crc, header_crc24));
+                    }
 
-                        let envelope = payload.split_to(envelope_len);
+                    let compressed_len = (header & 0x1ffff) as usize;
+                    let compressed_payload_end = compressed_len + COMPRESSED_FRAME_HEADER_LENGTH;
 
-                        tracing::debug!(
-                            "{}: incoming cassandra message:\n{}",
-                            self.direction,
-                            pretty_hex::pretty_hex(&envelope)
-                        );
+                    let frame_end = compressed_payload_end + FRAME_TRAILER_LENGTH;
 
-                        envelopes.push(Message::from_bytes(
-                            envelope.freeze(),
-                            crate::message::ProtocolType::Cassandra {
-                                compression: Compression::None,
-                            },
+                    let compressed_payload_crc32 = u32::from_le_bytes(
+                        frame_bytes[compressed_payload_end..frame_end]
+                            .try_into()
+                            .unwrap(),
+                    );
+
+                    let computed_crc =
+                        crc32(&frame_bytes[COMPRESSED_FRAME_HEADER_LENGTH..compressed_payload_end]);
+
+                    if compressed_payload_crc32 != computed_crc {
+                        return Err(payload_crc_mismatch_error(
+                            computed_crc,
+                            compressed_payload_crc32,
                         ));
                     }
 
+                    let self_contained = (header & (1 << 34)) != 0;
+                    if !self_contained {
+                        unimplemented!("Cannot support non-self contained frames yet");
+                    }
+
+                    let uncompressed_length = ((header >> 17) & 0x1ffff) as usize;
+
+                    frame_bytes.advance(COMPRESSED_FRAME_HEADER_LENGTH);
+
+                    let payload = if uncompressed_length == 0 {
+                        // protocol spec 2.2:
+                        // An uncompressed length of 0 signals that the compressed payload should be used as-is
+                        // and not decompressed.
+                        frame_bytes.split_to(compressed_len).freeze()
+                    } else {
+                        decompress(
+                            &frame_bytes.split_to(compressed_len).freeze(),
+                            uncompressed_length,
+                        )?
+                        .into()
+                    };
+
+                    let envelopes = self.extract_envelopes_from_payload(payload)?;
+
                     Ok(envelopes)
                 }
-                _ => unimplemented!(),
+                _ => Err(anyhow!("Only Lz4 compression is supported for v5")),
```

Comment:
We should do the same for the other "Only Lz4 compression is supported for v5" panics